src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -2702,11 +2769,51 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_DnsBrowse(const
 }
 
 template <class ImplClass>
-void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnDnsResolveResult(otError aError, const otDnsServiceResponse * aResponse,
-                                                                             void * aContext)
+void GenericThreadStackManagerImpl_OpenThread<ImplClass>::AddressResolveResult(otError aError,
+                                                                               const otDnsAddressResponse * aResponse,
+                                                                               void * aContext,
+                                                                               AsyncWorkFunct dispatchAddressResolve)
 {
     CHIP_ERROR error;
-    DnsResult * dnsResult = Platform::New<DnsResult>(aContext, MapOpenThreadError(aError));
+    DnsResult * dnsResult = reinterpret_cast<DnsResult *>(aContext);
+    otIp6Address address;
+
+    error = MapOpenThreadError(otDnsAddressResponseGetAddress(aResponse, 0, &address, nullptr));
```

Comment:
We may receive multiple addresses in the "Address Resolve" response. In the code here we seem to get and use the first one (index is `0`).

 If we later need to pass all (or select one if there are multiple), we can iterate through the here (keep reading different `index` values until `otDnsAddressResponseGetAddress()` returns `OT_ERROR_NOT_FOUND`.

Reply:
Yes, I know, however the structure used to represent DNS service in Matter currently includes only one IP address. Based on my local tests it seems the Matter devices also include only one IP address in DNS records, but it's a good point what should happen if there would be more of them. We can still pick only one IP address, but what criteria should we use to select it? @bzbarsky-apple do you have any opinion on that?

Reply:
@kkasperczyk-no The code that calls into this code expects to get a callback that contains:

1. The context it initially provided.
2. A `Dnssd::DnssdService *`.
3. A `Span<Inet::IPAddress>` of addresses.
4. The success-or-error status.

This code is choosing to keep track of that "list of addresses" in the single address slot of the `DnssdService`, but that's an implementation choice....  That said, refactoring this code to support the full list of addresses, if we want to do that, should be a separate PR.

Note, by the way, that in the "server includes AAAA with the SRV/TXT response" case you will only get a single address, based on what I see from the OpenThread APIs, not all the addresses.  So unless you do an unconditional separate resolve, you will just get whatever single address the DNS server picked there anyway.